scalajs and locales tests

cquiroz opened the pull request:
I'm sending this PR to discuss how to get `threetenbp` running in `scala.js`, filling some missing bits and adapting the tests. This PR is exploratory, it shouldn't be merged yet.

It includes
* Use a snapshot of the port of java locales: [scalajs-locales](https://github.com/cquiroz/scalajs-locales). For now, the `scalajs-locales` just supports  the `Locale` and `DecimalFormatSymbols` classes, and loading of data from CLDR.
* Convert `TestDecimalStyle` to junit
* Update the build to run the js tests using junit

Some important notes

- `TestDecimalStyle` was moved from `jvm/src/test` to `shared/src/test` to make it run in both environments JVM/JS
- For some reason I cannot access some `jUnit` functions that you can with the setup used in `scalajs-java-time` or `scalajs-locales`, in particular I get a linking error calling a function `assertEquals` where the arguments are primitives. I tried all kind of ways to get it to work, the dependencies are the same comparing projects, but still it breaks. So I replaced the `assertEquals` calls with `assertTrue(...)`, far from ideal
- I disabled the coverage tasks on travis, apparently they are not supported in scala.js
- `scalajs-locales` is still a work in progress, hence the snapshot, in the near future it should support the rest needed for `threetenbp`

Any comments are welcome. Do you think the project could be converted to be more like the `scalajs-java-time` project?

soc commented:
This looks pretty nice already!

> Any comments are welcome. Do you think the project could be converted to be more like the scalajs-java-time project?

Not sure what you have in mind ... what should we change?

cquiroz commented:
I found the issue removing the transitive jUnit instance brought by the novocode junit adapter. Now I can reuse the test code

cquiroz commented:
It seems novocode brings its own version of junit which was incompatbile, excluding it fixes the problem

```
"com.novocode" % "junit-interface" % "0.9" % "test" exclude("junit", "junit"))
```

soc commented on shared/src/test/scala/org/threeten/bp/format/TestDecimalStyle.scala:
```diff
@@ -31,14 +31,13 @@
 
 /** Test DecimalStyle. */
-@Test class TestDecimalStyle extends TestNGSuite {
+class TestDecimalStyle {
```
I'm a bit concerned about this. Won't this prevent the test from being picked up when running the tests on the JVM?

cquiroz replied:
Do you mean by removing the `@Test` annotation? It is very clear to mean that the test is executed in both JVM/JS. An easy way to test is to change one of the asserts to make it break. You see that it fails on both platforms

soc replied:
Ok, great!

soc replied:
I was also concerned about `TestNGSuite`, but that seems to come from ScalaTest, not TestNG itself.

sjrd commented on build.sbt:
```diff
@@ -24,18 +27,29 @@ lazy val threetenbpRoot = project.in(file("."))
+        "-v", "-a"),
+    libraryDependencies ++= Seq(
+      "com.github.cquiroz" %%% "scalajs-locales" % "0.1.0-SNAPSHOT",
+      "com.novocode" % "junit-interface" % "0.9" % "test" exclude("junit", "junit"))
```
You shouldn't even need this line at all. This is a JVM dependency, which should not appear on the classpath of a Scala.js project. `ScalaJSJUnitPlugin` already adds the `scalajs-junit-runtime` JS dependency, which includes the classes from `com.novocode.junit`.

cquiroz replied:
Thanks, I basically took that line from the build definition of scalajs-java-time
https://github.com/scala-js/scala-js-java-time/blob/master/build.sbt#L79
Indeed, it works without the `com.novocode` dependency

sjrd replied:
In scalajs-java-time, it is only in the *jvm*Settings ;)

sjrd commented on build.sbt:
```diff
@@ -24,18 +27,29 @@ lazy val threetenbpRoot = project.in(file("."))
+  .jsConfigure(_.enablePlugins(ScalaJSJUnitPlugin))
+
+lazy val threetenbp = threetenbpCross.jvm
+  .settings(commonSettings: _*)
```
Btw, you're entirely by-passing the intended use of `crossProject` here, and therefore you're missing all its benefits, i.e., not having to repeat the same things in the JS and JVM variants. You should write:
```scala
lazy val threetenbpCross = crossProject.crossType(CrossType.Full).in(file("."))
  .jsConfigure(_.enablePlugins(ScalaJSJUnitPlugin))
  .settings(commonSettings: _*)
  .jvmSettings(
    libraryDependencies ++= Seq(
      "org.scalatest" %%% "scalatest" % "3.0.0-M15" % "test",
      "org.testng" % "testng" % "6.9.10" % "test",
      "com.novocode" % "junit-interface" % "0.9" % "test"
    )
  )
  .jsSettings(
    resolvers +=
      "Sonatype OSS Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots",
    testOptions +=
      Tests.Argument(TestFramework("com.novocode.junit.JUnitFramework"),
        "-v", "-a"),
    libraryDependencies ++= Seq(
      "com.github.cquiroz" %%% "scalajs-locales" % "0.1.0-SNAPSHOT"
    )
  )

lazy val threetenbp = threetenbpCross.jvm
lazy val threetenbpJS = threetenbpCross.js
```

cquiroz replied:
Thanks for the comment, that indeed looks a better build definition

cquiroz commented:
Updated the PR with @srjd suggestion for the build code and a reference to the release version of `scala-java-locales`. It also rebased onto master and squashed.

soc commented on build.sbt:
```diff
@@ -27,13 +28,23 @@ lazy val threetenbpRoot = project.in(file("."))
+    )
+  ).jsSettings(
+    testOptions +=
+      Tests.Argument(TestFramework("com.novocode.junit.JUnitFramework"), "-v", "-a"),
```
Sorry if this is a stupid question: Is it necessary to set it here again? Shouldn't that be set from setting it in `settings(...)` above?

cquiroz replied:
You are right, is not needed here, we should remove it on the next PR

soc replied:
Ok, force pushed it ... I hope you didn't pull in the last few minutes ... :-)

soc commented:
Yay, I merged this in https://github.com/soc/scala-java-time/commit/4cff9065547eb31143cfc15ceddf5cbb0dca75c5

soc commented:
Ok, thanks!

I filed https://github.com/scala-js/scala-js/issues/2512 regarding the issue with `Arrays.asList`.
